subsys/bluetooth/host/gatt.c

```diff
@@ -373,7 +373,7 @@ static bool sc_ccc_cfg_write(struct bt_conn *conn,
 		}
 	}
 
-	return true;
+	return 0;
```

Comment:
Returning `ssize_t` implies that either you get a (negative) error or then you get the number of bytes written. Isn't it therefore wrong to return 0 here, since we acted on the full two bytes of the input (`u16_t value`)? Alternatively, if this API isn't intended to return number of bytes written you should change the return type from `ssize_t` to `int`.

Another thing thing that surprises me is that you didn't have to change any API documentation when doing this change, i.e. it seems this API was never documented? I think we need to fix that as well.

Reply:
OK, I will change return code to int. I guess that I also should change the .cfg_write callback to return a negative value in case of error using the BT_GATT_ERR() macro.
And yes, we miss the API documentation.

Reply:
@kimsekkelund Doesn't it make more sense to match the return value with the write callback, which has `ssize_t`, then we can just check for a valid positive number (1 or 2) otherwise we give an error.

@Vudentz Looking at the handling of the write callback I think there is an issue there with the return code. If you return `0` for example, you would get `BT_ATT_ERR_UNLIKELY` (since 0 is not valid), else you would get a random error code based on how much was written.
```
	if (write < 0 || write != data->len) {
		data->err = err_to_att(write);
		return BT_GATT_ITER_STOP;
	}
```

Reply:
> @Vudentz Looking at the handling of the write callback I think there is an issue there with the return code. If you return `0` for example, you would get `BT_ATT_ERR_UNLIKELY` (since 0 is not valid), ~else you would get a random error code based on how much was written.~ You would get `BT_ATT_ERR_UNLIKELY` for all.
> 
> ```
> 	if (write < 0 || write != data->len) {
> 		data->err = err_to_att(write);
> 		return BT_GATT_ITER_STOP;
> 	}
> ```

You mean if that would a zero length write so nothing really is written? But that would not cause an invalid behavior afaik since it is still not < 0 or different than the data->len, or you are saying 0 should always be translated to success, that is not really the case for the write callback.

Reply:
I'm saying that translating write to error code when the write is positive, but unequal to the length is strange. But maybe `BT_ATT_ERR_UNLIKELY` is good enough?

This makes more sense to me. (When I first made the comment I thought mistakenly that it would give some random error so maybe I'm making a bigger deal out of this than it actually is)
```
	if (write < 0 || write != data->len) {
		data->err = (write < 0) ? err_to_att(write)
					: BT_ATT_ERR_INVALID_ATTRIBUTE_LEN;
		return BT_GATT_ITER_STOP;
	}

```

Reply:
I don't see much difference either way since there is no error to inform something has been partially written, BT_ATT_ERR_UNLIKELY seems a little more generic though, in any case this is a local error with the application not indicating a proper error and not consuming all the data.

Reply:
Fair enough. Let's end that discussion then

Let's get back to the other issue of the return value of the callback. Doesn't it make sense to have it match `cfg_write` with the `write` callback?

Reply:
@joerchan I will align with the attribute write (returning ssize_t) - it makes sense.
This means that the cfg_write handler should return number of bytes written or an error code.